examples/find_if.cpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <unifex/just.hpp>
+#include <unifex/let.hpp>
+#include <unifex/scheduler_concepts.hpp>
+#include <unifex/sync_wait.hpp>
+#include <unifex/timed_single_thread_context.hpp>
+#include <unifex/find_if.hpp>
+#include <unifex/transform.hpp>
+#include <unifex/when_all.hpp>
+
+#include <chrono>
+#include <iostream>
+
+using namespace unifex;
+using namespace std::chrono;
+using namespace std::chrono_literals;
+
+int main() {
+  std::vector<int> input{1, 2, 3, 4};
+  // Apply linear find_if.
+  // As for std::find_if it returns the first instance that matches the
+  // predicate where the algorithm takes an iterator pair as the first
+  // two arguments, and forwards all other arguments to the predicate and
+  // onwards to the result.
+  // Precise API shape for the data being passed through is TBD, this is
+  // one option only.
+  std::optional<std::vector<int>::iterator> result = sync_wait(
+    transform(
+      find_if(
+          just(begin(input), end(input), 3),
```

Comment:
Ideally this would be taking a `range` as input, but I understand that's difficult at the moment given that `std::ranges` is not yet widely available in all standard libraries.

The other thing I'm vacillating over here is whether this algorithm should take a `sender_of<range>` or just the `range` as input.

The former might allow some better chaining ergonomics and optimisation possibilities, but makes it harder to use for simpler cases, e.g. from a coroutine, as you'd often need to wrap the input in a `just()` wrapper when you already have a range.

The latter would still support chaining, but would require something like the `let()` algorithm to connect the output of one sender to the argument to the function.
e.g. from a coroutine:
```
std::vector<int> v = co_await produce_some_vector();
auto it = co_await unifex::find_if(v, [](int i) { return i == 3; }, unifex::seq);
if (it != v.end()) {
  *it = 4;
  co_return true;
}
co_return false;
```
but from a pure sender approach:
```
let(
  produce_some_vector(),
  [](std::vector<int>& v) {
    return transform(
      find_if(v, [](int i) { return i == 3;}),
      [&](std::vector<int>::iterator it) {
        if (it != v.end()) {
          *it = 4;
          return true;
        }
        return false;
      });
  });
```
or with pizza-operator:
```
produce_some_vector()
|> let([](std::vector<int>& v) {
    return
      find_if(v, [](int i) { return i == 3;})
      |> transform([&](std::vector<int>::iterator it) {
        if (it != v.end()) {
          *it = 4;
          return true;
        }
        return false;
      });
  });
```


For now I think what you have is good enough to get started, we can refine the shape later.

Reply:
There are wider questions with this one.

Should we be able to pass a container rather than a view?
just(vector) | find_if(...) | what?

if we propagate an iterator at that point we've lost the container. So in that world. the let form seems necessary and if we have let why have it take a sender<view> rather than a view. What if we do want to perform a find on a data structure that is being generated, though?

Kirk suggested it should return some pair of a contained and an iterator, but then of course there are questions about correctness of the iterator on return of the container.